File: src/main/java/com/kiseru/asteroids/client/SimpleClient.java
package com.kiseru.asteroids.client;

import java.io.*;
import java.net.Socket;

public class SimpleClient {
    private static final int PORT = 6501;
    private static final String HOST = "localhost";

    public static void main(String[] args) throws IOException {
        try (Socket s = new Socket(HOST, PORT)) {
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(s.getInputStream()));
            PrintWriter writer = new PrintWriter(s.getOutputStream(), true);
            Thread receiver = new Thread(new Receiver(bufferedReader));
            receiver.start();
            Thread sender = new Thread(new Sender(writer));
            sender.start();
            receiver.join();
            sender.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}

File: src/main/kotlin/com/kiseru/asteroids/client/SimpleClient.kt
package com.kiseru.asteroids.client

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.ConnectException
import java.net.Socket

private const val PORT = 6501
private const val HOST = "localhost"

suspend fun main(): Unit =
    coroutineScope {
        val socket = connectToServer()
        launch(Dispatchers.IO) {
            val listener = ServerListener(socket.getInputStream(), System.out)
            listener.listen()
        }

        launch(Dispatchers.IO) {
            val client = ServerClient(System.`in`, socket.getOutputStream())
            client.startClient()
        }
    }

private suspend fun connectToServer(): Socket {
    while (true) {
        try {
            println("Trying to connect to server...")
            return withContext(Dispatchers.IO) { Socket(HOST, PORT) }
        } catch (e: ConnectException) {
            println("Failed to connect to server. Next attempt in 10 seconds.")
            delay(10000)
        }
    }
}
